source/agora/consensus/validation/GenesisBlock.d

```diff
@@ -0,0 +1,216 @@
+}
+
+/*******************************************************************************
+
+    Check the validity of a genesisBlock.
+
+    A genesisBlock is considered valid if:
+        - Block height is 0
+        - The previous block hash of the block must be empty
+        - The number of transactions in the block is above zero and less than
+          Block.TxsInBlock
+        - Transactions are ordered by their hash value
+        - All the transactions pass validation, which implies:
+            - Transaction type in block should only be Payment or Freeze
+            - There is only one input for the transaction
+            - The signature value of the input of the transaction is genuine
+            - There is more than one output of the transaction
+            - The output of the transaction should not be empty at address
+            - The output.value of the transaction must be in accordance
+              with the `Amount` rule
+        - The merkle root in the header matches the re-built merkle tree root
+        - All the enrollments pass validation, which implies:
+            - The Enrolments should be arranged in ascending order by utxo_key
+            - Signature are authentic
+            - The enrollments refer to freeze tx's in this block
+            - The freeze tx's spend an output which must be found with
+              the findUTXO() delegate
+            - Freeze transactions referenced in registration should not be
+              double-spend
+
+    Params:
+        block = The genesisBlock to check
+
+    Returns:
+        `null` if the genesisBlock is valid, a string explaining the reason it
+        is invalid otherwise.
+
+*******************************************************************************/
+
+public string isInvalidReason (const ref Block genesisBlock) nothrow @safe
+{
+    if (genesisBlock == Block.init)
+        return "GenesisBlock: The block is empty";
+
+    if (genesisBlock.header.height != 0)
+        return "GenesisBlock: The height of the block is not 0";
+
+    if (genesisBlock.header.prev_block != Hash.init)
+        return "GenesisBlock: Header.prev_block is not empty";
+
+    if (genesisBlock.txs.length == 0)
+        return "GenesisBlock: Transaction(s) is empty";
+
+    if (!genesisBlock.txs.isSorted())
+        return "GenesisBlock: Transactions are not sorted";
+
+    if (genesisBlock.txs.length >= Block.TxsInBlock)
+        return "GenesisBlock: The number of transactions is out of bounds";
+
+    UTXOSetValue[Hash] utxo_set;
+    foreach (const ref tx; genesisBlock.txs)
+    {
+        if (!(tx.type == TxType.Payment || tx.type == TxType.Freeze))
+            return "GenesisBlock: Transaction type allows only "
+                ~ "Payment and Freeze";
+
+        if (tx.inputs.length == 0)
+            return "GenesisBlock: No a input in the transaction";
+
+        if (tx.inputs.length != 1)
+            return "GenesisBlock: Transaction must have only one input";
```

Comment:
You just need one check, not both.

And, you can decide to change the rule here. Make inputs equal zero rather than one. There is no real input.